File: solutions/python/math/factorial.py
#!/usr/bin/env python3
"""
Factorial
Calculate factorial of a number using recursion and iteration
Category: math | Difficulty: Easy
Date: 2026-03-07 | Author: Jay Singh
"""

def factorial_recursive(n):
    """Calculate factorial using recursion. Time: O(n), Space: O(n)"""
    if n <= 1:
        return 1
    return n * factorial_recursive(n - 1)

def factorial_iterative(n):
    """Calculate factorial using iteration. Time: O(n), Space: O(1)"""
    result = 1
    for i in range(2, n + 1):
        result *= i
    return result

if __name__ == "__main__":
    for num in [0, 1, 5, 10, 15]:
        print(f"{num}! = {factorial_recursive(num)} (recursive)")
        print(f"{num}! = {factorial_iterative(num)} (iterative)")

File: solutions/c/math/factorial.c
/*
 * Factorial
 * Calculate factorial of a number using recursion and iteration
 * Category: math | Difficulty: Easy
 * Date: 2026-03-12 | Author: Jay Singh
 */

#include <stdio.h>
#include <stdlib.h>

// Time: O(n), Space: O(1)
long long factorial_recursive(int n) {
    if (n <= 1) return 1;
    return n * factorial_recursive(n - 1);
}

long long factorial_iterative(int n) {
    long long result = 1;
    for (int i = 2; i <= n; i++) result *= i;
    return result;
}

int main() {
    int nums[] = {0, 1, 5, 10, 15, 20};
    for (int i = 0; i < 6; i++) {
        printf("%d! = %lld (recursive)\n", nums[i], factorial_recursive(nums[i]));
        printf("%d! = %lld (iterative)\n", nums[i], factorial_iterative(nums[i]));
    }
    return 0;
}
